### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by asampal** [mads-hartmann/bash-language-server#5]
Seeing this when I switch to a bash file, in Visual Studio Code 1.20.1, Windows 10, extension v. 1.1.4:

```
module.js:672
  return process.dlopen(module, path._makeLong(filename));
                 ^

Error: \\?\C:\Users\asampaleanu\.vscode\extensions\mads-hartmann.bash-ide-vscode-1.1.4\server\node_modules\tree-sitter\build\Release\tree_sitter_runtime_binding.node is not a valid Win32 application\\?\C:\Users\asampaleanu\.vscode\extensions\mads-hartmann.bash-ide-vscode-1.1.4\server\node_modules\tree-sitter\build\Release\tree_sitter_runtime_binding.node
    at Object.Module._extensions..node (module.js:672:18)
    at Module.load (module.js:556:32)
    at tryModuleLoad (module.js:499:12)
    at Function.Module._load (module.js:491:3)
    at Module.require (module.js:587:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (C:\Users\asampaleanu\.vscode\extensions\mads-hartmann.bash-ide-vscode-1.1.4\server\node_modules\tree-sitter\index.js:1:125)
    at Module._compile (module.js:643:30)
    at Object.Module._extensions..js (module.js:654:10)
    at Module.load (module.js:556:32)
[Info  - 9:28:54 AM] Connection to server got closed. Server will restart.
module.js:672
  return process.dlopen(module, path._makeLong(filename));
```

**Comment by mads-hartmann** [comment]
@asampal Thanks, I believe I know how to fix this. I've accidentally shipped an OS X binary I believe ;)

**Comment by asampal** [comment]
Great - looking forward to the fix.

**Comment by mads-hartmann** [comment]
@asampal Could you try installing version `1.1.6`? I believe I've fixed the issue ☺️

**Comment by mads-hartmann** [comment]
@asampal Please reopen if you're still having issues 🙂

**Comment by asampal** [comment]
I'm still seeing the same problem @mads-hartmann .  Should the Windows binary be built automatically when the extension is installed? The error message is exactly the same, so it looks like the extension is still somehow not building for Windows.

**Comment by asampal** [comment]
Could it be that my system is somehow being detected as running OS X?

**Comment by mads-hartmann** [comment]
@asampal Turns out it's a bit more complicated to ship an extension that has native dependencies than I thought 🙂I found this [guide](https://github.com/electron/electron/blob/master/docs/tutorial/using-native-node-modules.md) that I'll try to follow.

**Comment by mads-hartmann** [comment]
@asampal I'm trying a different approach. Can you run `npm i -g bash-language-server` for me to see if the server installs on your system. And if it does, can you then try running `bash-language-server -h` to see if it executes properly?

**Comment by asampal** [comment]
I get this when I try to install the language server:


```
adi@VESUVIUS C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter                                                                                                                                
$ if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files\nodejs\node_modules\n
pm\node_modules\node-gyp\bin\node-gyp.js" rebuild )                                                                                                                                                                                     
Warning: Missing input files:                                                                                                                                                                                                           
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\prepare_grammar\expand_tokens.cc                                          
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\syntax_grammar.cc                                                         
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\prepare_grammar\token_description.cc                                      
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\build_tables\lookahead_set.cc                                             
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\build_tables\lex_item_transitions.cc                                      
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\prepare_grammar\expand_repeats.cc                                         
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\parse_table.cc                                                            
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\externals\json-parser\json.c                                                           
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\prepare_grammar\extract_choices.cc                                        
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\prepare_grammar\normalize_rules.cc                                        
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\rules\character_set.cc                                                    
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\generate_code\c_code.cc                                                   
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\prepare_grammar\extract_tokens.cc                                         
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\prepare_grammar\flatten_grammar.cc                                        
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\prepare_grammar\intern_symbols.cc                                         
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\rules\choice.cc                                                           
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\prepare_grammar\prepare_grammar.cc                                        
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\build_tables\lex_table_builder.cc                                         
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\rules\metadata.cc                                                         
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\rule.cc                                                                   
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\lex_table.cc                                                              
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\build_tables\lex_item.cc                                                  
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\build_tables\parse_table_builder.cc                                       
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\build_tables\parse_item.cc                                                
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\compile.cc                                                                
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\build_tables\parse_item_set_builder.cc                                    
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\rules\repeat.cc                                                           
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\rules\seq.cc                                                              
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\prepare_grammar\parse_regex.cc                                            
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\build_tables\rule_can_be_blank.cc                                         
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\util\string_helpers.cc                                                    
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\precedence_range.cc                                                       
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\build_tables\lex_conflict_manager.cc                                      
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\..\..\..\vendor\tree-sitter\src\compiler\parse_grammar.cc                                                          
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.                                                                                                                             
C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter\build\vendor\tree-sitter\runtime.vcxproj(28,3): error MSB4019: The imported project "C:\Microsoft.Cpp.Default.props" was not found. Confirm 
 that the path in the <Import> declaration is correct, and that the file exists on disk.                                                                                                                                                
gyp ERR! build error                                                                                                                                                                                                                    
gyp ERR! stack Error: `C:\Windows\Microsoft.NET\Framework\v4.0.30319\msbuild.exe` failed with exit code: 1                                                                                                                              
gyp ERR! stack     at ChildProcess.onExit (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:258:23)                                                                                                          
gyp ERR! stack     at ChildProcess.emit (events.js:127:13)                                                                                                                                                                              
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:209:12)                                                                                                                                            
gyp ERR! System Windows_NT 10.0.16299                                                                                                                                                                                                   
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"                                                                               
gyp ERR! cwd C:\Users\adi\AppData\Roaming\npm\node_modules\bash-language-server\node_modules\tree-sitter                                                                                                                                
gyp ERR! node -v v9.6.1                                                                                                                                                                                                                 
gyp ERR! node-gyp -v v3.6.2                                                                                                                                                                                                             
gyp ERR! not ok                                                                                                                                                                                                                         
npm ERR! code ELIFECYCLE                                                                                                                                                                                                                
npm ERR! errno 1                                                                                                                                                                                                                        
npm ERR! tree-sitter@0.9.2 install: `node-gyp rebuild`                                                                                                                                                                                  
npm ERR! Exit status 1                                                                                                                                                                                                                  
npm ERR!                                                                                                                                                                                                                                
npm ERR! Failed at the tree-sitter@0.9.2 install script.                                                                                                                                                                                
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.                                                                                                                                      
                                                                                                                                                                                                                                        
npm ERR! A complete log of this run can be found in:                                                                                                                                                                                    
npm ERR!     C:\Users\adi\AppData\Roaming\npm-cache\_logs\2018-03-03T21_03_30_599Z-debug.log                                                                                                                                            
```

**Comment by asampal** [comment]
The full log is for the above is:

[2018-03-03T21_03_30_599Z-debug.log](https://github.com/mads-hartmann/bash-language-server/files/1777899/2018-03-03T21_03_30_599Z-debug.log)

**Comment by asampal** [comment]
Maybe you could, at least as an option, use the WSL (Windows Subsystem for Linux) for what you need to do. Since the Fall Creators Update it's quite interoperable with Windows. Some of the files [here](https://github.com/watzon/wsl-proxy) might be of help.

**Comment by mads-hartmann** [comment]
@asampal I'd love your help on this. I don't have access to a windows computer right now so it's hard for me to work on it and my knowledge of windows is very limited.

**Comment by mads-hartmann** [comment]
The goal is to get `npm i -g bash-language-server` to work on Windows. If that works the rest of the client then we should have windows support as the client is platform independent.

**Comment by seflue** [comment]
After playing around with various settings of msvs_version I got bash-language-server installed via npm on Windows 10. But now I get the following error message in VSC: 'Can't find bash-langauge-server [sic!] on your PATH. Please install it using `npm i -g bash-language-server`'. But looking in my environment variables I found the relevant folder (`C:\Users\myuser\AppData\Roaming\npm`) in the Path variable.

**Comment by mads-hartmann** [comment]
@seflue Thanks for trying it out on Windows. You can see [here](https://github.com/mads-hartmann/bash-language-server/blob/master/vscode-client/src/util.ts#L19) how it tries to figure out where `bash-language-server` is installed. I'll see if I can add a settings where you can explicitly set the path to the `bash-language-server` executable so it's a bit easier to debug :)

**Comment by seflue** [comment]
@mads-hartmann: Could you try both cases of command, `bash-language-server` and `bash-language-server.cmd` when creating the promise? The latter is the command for Windows and it needs the `.cmd`-Suffix for recognizing the script type, because ... Windows, you know? ;-) The first is the bash script used by Linux.

**Comment by mads-hartmann** [comment]
@seflue Definitely. I should be able to detect if you're on windows or not and then use the `cmd` extension. If you're up for it you could try to work on the extension locally and use the `.cmd` extension to see if everything works then :)

**Comment by mads-hartmann** [comment]
@seflue I've published version `1.2.1` which adds `.cmd` if you're running on windows. I'd love if you could give it a go :)

**Comment by seflue** [comment]
@mads-hartmann: I tried the new version, but I still get the same error. It would be nice if you can write out some error log, to check, what is going wrong when the plugin tries to execute the command.

**Comment by mads-hartmann** [comment]
@seflue Thanks for trying :) I'll add some more logging. Could you explain how you managed to install it on Windows? That would make it possible for me to debug it :)

**Comment by seflue** [comment]
@mads-hartmann: I can't give you an exact recipe, but I fiddled around a little bit with the tips on the Internet. I ran `npm install --global --production windows-build-tools` on an administrator powershell and changed the `msvs_version` to 2015 with `npm config set msvs_version 2015 --global` also with administrator permissions. After some trying with these commands under different permissions (user and administrator) I was able to install bash-language-server successfully with `npm i -g bash-language-server`. It broke my BuildTools 2017 setup and cost me some time to get it up and running again, but I managed it somehow.

**Comment by henson-dev** [comment]
Here's what I get:

C:\WINDOWS\system32>bash-language-server
Unknown command 'C:\Program Files\nodejs\node.exe C:\Users\User\AppData\Roaming\npm\node_modules\bash-language-server\bin\main.js'. Run with -h for help.

C:\WINDOWS\system32>bash-language-server.cmd
Unknown command 'C:\Program Files\nodejs\node.exe C:\Users\User\AppData\Roaming\npm\node_modules\bash-language-server\bin\main.js'. Run with -h for help.

Could you somehow be quoting the path of the program and the argument before calling it?

**Comment by PromoFaux** [comment]
@mads-hartmann 

>I've published version 1.2.1 which adds .cmd 

I may or may not be doing something wrong. Was unable to get it intsall on windows, so have installed on WSL and used the `wsl-proxy` method you mentioned a few comments up... VSCode still complaining that it can't find `bash-language-server` on my `PATH`.. am I doing something wrong that is glaringly obvious?? :)

![image](https://user-images.githubusercontent.com/1998970/39085378-5b878526-4579-11e8-855d-3745106f20c1.png)

![image](https://user-images.githubusercontent.com/1998970/39085407-ba938be6-4579-11e8-8ab5-e46ca3013de0.png)

![image](https://user-images.githubusercontent.com/1998970/39085411-ca3eec8e-4579-11e8-90b8-d1ab3e85a336.png)

**Comment by PromoFaux** [comment]
Worth noting it works from `cmd`

![image](https://user-images.githubusercontent.com/1998970/39085442-3c1747e8-457a-11e8-942f-e33c60cee106.png)

**Comment by PromoFaux** [comment]
Getting somewhere, `getBasePath()` is falling over. Looks as though using the wsl-proxy doesn't work/isn't compatible. I'll keep digging.

![image](https://user-images.githubusercontent.com/1998970/39085877-7ed9269e-4581-11e8-8b8e-1077eefe8049.png)

**Comment by PromoFaux** [comment]
Lots of tinkering this afternoon, but didn't really get anywhere with it. It appears that attempting to run `bash-language-server` via the batch file that redirects it to `WSL` does not work too great.

I did, however, discover that you can get it working under WSL by using the Linux version of Code via an Xserver.

Bit of a hacky workaround I guess, but I'm satisfied with the outcome. 

TL;DR - Compile `bash-language-server` on WSL, get [linux Code working via WSL](https://github.com/Microsoft/WSL/issues/2760), this now "works on windows" 😉

**Comment by mads-hartmann** [comment]
@PromoFaux Thanks for documenting your attempts :) I hope it will help other people to get it to work on Windows, even if it is using WSL 👍

**Comment by rickywu** [comment]
should I use 'sh': ['bash-language-server', 'start']
or  'sh': ['bash-language-server.cmd', 'start']

bash-language-server start works well in cmd but not work when use in Vim

**Comment by beltaurus** [comment]
I was able to get this to work on Windows 10 through Remote WSL without any issues.
However, I'm can't really go to the newer versions, since I'm doing everything from the Ubuntu18.04LTS repository (which doesn't get the latest versions of Bash IDE, or shellchecker).

I was able to install a newer version of shellchecker as a "native" Windows 10 application through "chocolatey", giving me hope to run Bash IDE without WSL as well.

However, the "recipe" to build the bash-language-server doesn't seem to work for me...

1. I installed node.js (which includes npm) through chocolatey (https://chocolatey.org/packages/nodejs.install).
2. I used npm in a CMD shell with administration privileges to install windows-build-tools (npm i -g windows-build-tools); Note that I have Microsoft Visual Studio VS2017 Community installed, I'm not sure if this interferes.
3. I noticed that this windows-build-tools package also installs Python2, which is fine with me...
4. I tried to install: npm i -g bash-language-server; I tried to set my msvc_version to 2015 or 2017, but both attempts result in build errors (related to a mismatch of expected arguments versus actual arguments in library functions):
![image](https://user-images.githubusercontent.com/54114043/64548023-aee20280-d2e2-11e9-8f3e-065a139d7089.png)

Before I try to detangle that any further, can we confirm that this is the correct approach to get a "native" version of bash-language-server to work on Windows 10?

**Comment by beltaurus** [comment]
Since this is still open, and there doesn't seem to be a clear recipe confirmed:
1. I experimented with different versions of NodeJS, and different distributors (e.g. https://chocolatey.org/packages/nodejs); However, the _only_ one that led to a working solution is https://nodejs.org/en/, where I downloaded the installer for 10.16.3LTS. When I installed it, I made sure to select "Add to PATH", for easy access of the "npm" command.
![image](https://user-images.githubusercontent.com/54114043/64574181-45ceaf00-d323-11e9-9e02-2aac444a7a47.png).
This installation went very easy, and there were no "dependencies" throwing in any monkey wrenches.

2. I went ahead and launched a CMD-console with Administration privileges, and dared to execute:
`npm i -g bash-language server
`
Again extremely smooth (any other version of NodeJS that I tried was insisting on installing Python2, installing Windows Build packages,... I had cleared those from my system, and this npm-based installation was able to create the bash-language-server without complaints...
![image](https://user-images.githubusercontent.com/54114043/64574352-e7560080-d323-11e9-85af-ce42e9b58896.png)

3. In VS Code, I was now ready to install the Local User Extension Bash ID. The only thing to pay attention to is the Bash ID setting for the Path:
![image](https://user-images.githubusercontent.com/54114043/64574465-7236fb00-d324-11e9-89e9-d58eff5ed2f1.png)

If you use the command with the '.cmd' suffix, you're all set! If you use the one with out the suffix, you'll get the dreaded:
![image](https://user-images.githubusercontent.com/54114043/64574490-8f6bc980-d324-11e9-92bf-93355957944f.png)


And I'm now using it locally, happily ever after.

**Comment by gerwinkramer** [comment]
Thanks @beltaurus ! I'm using NVM For Windows so I had to set the "Bash Ide: Path" setting to something like: C:\Users\MyUserName\AppData\Roaming\nvm\v10.16.3\bash-language-server.cmd (note "nvm\v10.16.3" instead of npm in the path) and now it works 👍

**Comment by skovhus** [comment]
With the new version (1.4.0) of the VS Code extension, we got rid of the manual npm install step of the bash-language-server. Hopefully, this will resolve npm installation issues. :) 

The latest version of `bash-language-server` (the npm package) should also be easier to install.

Let me know if this solves the issues.
